lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -253,51 +435,360 @@ void FileEmitter::emitInstance(llvm::json::OStream &json,
+          .Case<comb::SubOp>([](auto) { return "-"; })
+          .Case<comb::MulOp>([](auto) { return "*"; })
+          .Case<comb::DivUOp, comb::DivSOp>([](auto) { return "/"; })
+          .Case<comb::ModUOp, comb::ModSOp>([](auto) { return "%"; })
+          .Case<comb::ShlOp>([](auto) { return "<<"; })
+          .Case<comb::ShrUOp>([](auto) { return ">>"; })
+          .Case<comb::ShrSOp>([](auto) { return ">>>"; })
+          .Case<comb::ICmpOp>([](auto cmpOp) -> StringRef {
+            switch (cmpOp.getPredicate()) {
+            case comb::ICmpPredicate::eq:
+              return "==";
+            case comb::ICmpPredicate::ne:
+              return "!=";
+            case comb::ICmpPredicate::ceq:
+              return "===";
+            case comb::ICmpPredicate::cne:
+              return "!==";
+            case comb::ICmpPredicate::weq:
+              return "==?";
+            case comb::ICmpPredicate::wne:
+              return "!=?";
+            case comb::ICmpPredicate::ult:
+            case comb::ICmpPredicate::slt:
+              return "<";
+            case comb::ICmpPredicate::ugt:
+            case comb::ICmpPredicate::sgt:
+              return ">";
+            case comb::ICmpPredicate::ule:
+            case comb::ICmpPredicate::sle:
+              return "<=";
+            case comb::ICmpPredicate::uge:
+            case comb::ICmpPredicate::sge:
+              return ">=";
+            }
+            return {};
+          })
+          .Default([](auto) { return ""; });
+  if (!binaryOpcode.empty() && op->getNumOperands() == 2) {
+    auto lhs = emitExpression(op->getOperand(0));
+    auto rhs = emitExpression(op->getOperand(1));
+    if (!lhs || !rhs)
+      return {};
+    return {JObject({
+                {"opcode", binaryOpcode},
+                {"operands", {lhs.expr, rhs.expr}},
+            }),
+            result.getType()};
+  }
+
+  // Expand variadic combinational ops into nested binary HGLDD operations.
+  StringRef variadicOpcode = TypeSwitch<Operation *, StringRef>(op)
+                                 .Case<comb::AndOp>([](auto) { return "&"; })
+                                 .Case<comb::OrOp>([](auto) { return "|"; })
+                                 .Case<comb::XorOp>([](auto) { return "^"; })
+                                 .Default([](auto) { return ""; });
+  if (!variadicOpcode.empty()) {
+    auto operands = op->getOperands();
+    auto value = emitExpression(operands[0]);
+    if (!value)
+      return {};
+    operands = operands.drop_front();
+    while (!operands.empty()) {
+      auto otherValue = emitExpression(operands[0]);
+      if (!otherValue)
+        return {};
+      operands = operands.drop_front();
+      value = {JObject({
+                   {"opcode", variadicOpcode},
+                   {"operands", {value.expr, otherValue.expr}},
+               }),
+               result.getType()};
     }
+    return value;
```

Comment:
Does HGLDD have to legalize these variadic operations into binary operations?

Reply:
Yes unfortunately. The expressions we can put into HGLDD are fairly straightforward, and fairly close to the Verilog AST. It expects unary and binary operations, no variadics.

Reply:
Oh sorry, upon a second read I think I understand: you mean that after ExportVerilog all variadic comb ops have been legalized into binary comb ops, so this emitter no longer has to deal with variadic ops. Is that accurate?

Reply:
Yes ExportVerilog lowers variadic operations except for concat op into binary operations